californium-core/src/main/java/org/eclipse/californium/core/network/MessageExchangeStore.java

```diff
@@ -101,17 +104,22 @@
 	 * Removes the exchange registered under a given token.
 	 * 
 	 * @param token the token of the exchange to remove.
+	 * @param exchange Exchange to be removed, if registered with provided token.
 	 */
-	void remove(KeyToken token);
+	void remove(KeyToken token, Exchange exchange);
 
 	/**
 	 * Removes the exchange registered under a given message ID.
 	 * 
 	 * @param messageId the message ID to remove the exchange for.
+	 * @param exchange Exchange to be removed. If {@code null}, the current
+	 *                 exchange with the MID is removed. If not {@code null},
+	 *                 only this exchange is removed, if it's registered with
+	 *                 the MID.
 	 * @return the exchange that was registered under the given message ID or {@code null}
 	 * if no exchange was registered under the ID.
 	 */
-	Exchange remove(KeyMID messageId);
+	Exchange remove(KeyMID messageId, Exchange exchange);
```

Comment:
The docs for the returned value need to be updated with the additional constraint introduced by the *exchange* param, shouldn't it?